File: cpp/examples/linux/pressure/BMP384/complete/main.cpp
#ifndef TEST_I2C_BUS
#define TEST_I2C_BUS 1
#endif
#ifndef TEST_ADDR
#define TEST_ADDR 0x76
#endif

#include <cstdio>
#include <cstring>
#include "I2CConnectionLinux.h"
#include "BMP384.h"

static int passed = 0, failed = 0;

static void check_true(bool cond, const char *label) {
    if (cond) { printf("PASS %s\n", label); passed++; }
    else       { printf("FAIL %s\n", label); failed++; }
}

int main() {
    I2CConnectionLinux connection(TEST_I2C_BUS, TEST_ADDR);
    BMP384Full bmp(connection, /*spi=*/false);             // Create BMP384 driver, (connection, spi=false)

    bmp.configure(4, 1, 2, 0x03);                          // Configure ADC and IIR filter, (osr_p 0–5, osr_t 0–5, iir_filter 0–7, odr_sel 0x00–0x11) → None
                                                            // sets oversampling, IIR coefficient, and output data rate
    bmp.set_mode(BMP384Full.MODE_NORMAL);                   // Set power mode, (mode 0/1/3) → None
    bool ready = bmp.is_data_ready();                       // Check data-ready flag, () → bool
                                                            // true if STATUS.drdy_press is set
    float t = bmp.temperature();                            // Read temperature, () → float °C
    float p = bmp.pressure();                               // Read pressure, () → float hPa
    float tp = 0.0f, tt = 0.0f;
    bmp.read(tp, tt);                                       // Read both values in one burst, (pressure_hpa, temperature_c) → None
    float fp = 0.0f, ft = 0.0f;
    bmp.read_forced(fp, ft);                                // Trigger forced measurement and read, (pressure_hpa, temperature_c) → None
    bmp.fifo_configure(true, true, 64);                     // Configure FIFO, (press_en bool, temp_en bool, wtm 0–511, stop_on_full=false) → None
                                                            // enables FIFO, sets watermark, arms pressure+temperature frames
    const char* types[16];
    double values[16];
    size_t n = bmp.fifo_read(types, values, 16);            // Read and parse FIFO frames, (types, values, max_frames) → size_t
    bmp.fifo_flush();                                       // Flush FIFO contents, () → None
    float alt = bmp.altitude();                             // Compute altitude, (sea_level_hpa=1013.25) → float m
    bmp.softreset();                                        // Soft reset chip, () → None

    check_true(t > -40.0f && t < 85.0f, "temperature_in_range");
    check_true(p > 300.0f && p < 1250.0f, "pressure_in_range");
    printf("T=%.1f C, P=%.1f hPa, ready=%d, frames=%d, alt=%.1f m\n",
        t, p, ready, (int)n, alt);

    printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}

File: rust/examples/linux/pressure/bmp384/complete/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::pressure::{Bmp384Full, MODE_NORMAL, MODE_FORCED};

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x76);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut bmp = Bmp384Full::new(dev, addr, false).expect("init BMP384"); // Create BMP384 driver, (i2c, addr=0x76)

    bmp.configure(4, 1, 2, 0x03).expect("configure");             // Configure ADC and IIR filter, (osr_p 0–5, osr_t 0–5, iir_filter 0–7, odr_sel 0x00–0x11) → Result<()>
                                                                    // sets oversampling, IIR coefficient, and output data rate
    bmp.set_mode(MODE_NORMAL).expect("set_mode");                  // Set power mode, (mode) → Result<()>
    let ready = bmp.is_data_ready().expect("is_data_ready");       // Check data-ready flag, () → Result<bool>
                                                                    // true if STATUS.drdy_press is set
    let t = bmp.temperature().expect("temperature");               // Read temperature, () → f32 °C
    let p = bmp.pressure().expect("pressure");                     // Read pressure, () → f32 hPa
    let (_p, _t) = bmp.read().expect("read");                      // Read both values in one burst, () → (f32, f32)
    let (_p2, _t2) = bmp.read_forced().expect("read_forced");      // Trigger forced measurement and read, () → (f32, f32)
    bmp.fifo_configure(true, true, 64, false).expect("fifo_configure"); // Configure FIFO, (press_en bool, temp_en bool, wtm u16, stop_on_full bool) → Result<()>
                                                                    // enables FIFO, sets watermark, arms pressure+temperature frames
    let mut frames: [periph::chips::pressure::Bmp384FifoFrame; 16] = unsafe { std::mem::zeroed() };
    let _n = bmp.fifo_read(&mut frames).expect("fifo_read");       // Read and parse FIFO frames, (&mut [Bmp384FifoFrame]) → Result<usize>
    bmp.fifo_flush().expect("fifo_flush");                         // Flush FIFO contents, () → Result<()>
    let alt = bmp.altitude(1013.25).expect("altitude");            // Compute altitude, (sea_level_hpa f32) → f32 m
                                                                    // uses barometric formula to convert pressure to metres
    bmp.softreset().expect("softreset");                           // Soft reset chip, () → Result<()>
                                                                    // writes 0xB6 to CMD, waits 2 ms, re-reads calibration

    println!("T={:.1} C, P={:.1} hPa, ready={}, alt={:.1} m", t, p, ready, alt);
    // Avoid unused-variable warning for MODE_FORCED reference.
    let _ = MODE_FORCED;
}
